### PR TITLE
NMS-13479: Reload classification decision tree asynchronously in DefaultClassificationService

### PR DESCRIPTION
**Pull request by swachter**
Issue: https://issues.opennms.org/browse/NMS-13479

While editing classification rules, reloads of the classification engine are triggered repeatedly. A reload may take a couple of seconds depending on rule set sizes. In order to avoid sluggish behavior, reloading is done asynchronously. Only when the classification engine is used calls are blocked until any ongoing reloads did finish.

These changes fix the tests in `ClassificationRulePageIT`.

### CODE REVIEW

File: smoke-test/src/test/java/org/opennms/smoketest/ClassificationRulePageIT.java [hunk 1]
```diff
@@ -477,10 +480,20 @@ public String classify(ClassificationRequestDTO classificationRequest) {
...
-            sleep(DEFAULT_WAIT_TIME);
 
-            // Fiddle result out of UI
-            return execute(() -> findElementById("classification-response")).getText();
+            String textAfterClassification = "";
+
+            // wait until the classification response value change to a non-empty string that is different from the value
+            // before classification was started
+            for (int i = 0; i < 30; i++) {
+                sleep(DEFAULT_WAIT_TIME);
```
Comment (cgorantla) [inline smoke-test/src/test/java/org/opennms/smoketest/ClassificationRulePageIT.java:489]: You can use  `await().atMost()` pattern here instead of sleep. https://github.com/awaitility/awaitility

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java [hunk 2]
```diff
@@ -358,4 +369,124 @@ private void checkForDuplicateName (Group group) {
...
+                }
+            }
+        }
+
+        private void doReload() {
+            // this method must not modify the state because it not synchronized
+            try {
+                LOG.debug("reload classification engine");
+                delegate.reload();
+                if (reloadSucceeded()) {
```
[the remark this answers is not in the document]
Comment (fooker) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:428]: Yeah, this is a little bit confusing. Second interpretation could be that these methods are event handlers or state change handlers. Then they should be called `onReloadSucceeded()`, etc.

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationEngine.java [hunk 3]
```diff
@@ -110,9 +110,9 @@ public void reload() {
                     .append("minComp  : " + tree.info.minComp).append('\n')
                     .append("maxComp  : " + tree.info.maxComp).append('\n')
                     .append("avgComp  : " + (double) tree.info.sumComp / tree.info.leaves).append('\n')
-                    .append("minRuleSetSize :" + tree.info.minRuleSetSize).append('\n')
-                    .append("maxRuleSetSize :" + tree.info.maxRuleSetSize).append('\n')
-                    .append("avgRuleSetSize :" + (double) tree.info.sumRuleSetSize / tree.info.leaves).append('\n');
+                    .append("minLeafSize : " + tree.info.minLeafSize).append('\n')
```
Comment (fooker) [inline features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationEngine.java:113]: Unnecessary refactoring. Please extract in single commit next time (read about `git add -p -i` to split these changes in multiple commits).

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java [hunk 4]
```diff
@@ -79,11 +87,14 @@ public DefaultClassificationService(ClassificationRuleDao classificationRuleDao,
                                         SessionUtils sessionUtils) {
         this.classificationRuleDao = Objects.requireNonNull(classificationRuleDao);
         this.classificationGroupDao = Objects.requireNonNull(classificationGroupDao);
-        this.classificationEngine = Objects.requireNonNull(classificationEngine);
+        this.classificationEngine = new AsyncReloadingClassificationEngine(Objects.requireNonNull(classificationEngine));
```
Comment (fooker) [inline features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:90]: I think the wrapping should be done in the application context wiring.

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java [hunk 5]
```diff
@@ -79,11 +87,14 @@ public DefaultClassificationService(ClassificationRuleDao classificationRuleDao,
...
         this.sessionUtils = Objects.requireNonNull(sessionUtils);
         this.ruleValidator = new RuleValidator(filterService);
         this.groupValidator = new GroupValidator(classificationRuleDao);
         this.csvService = new CsvServiceImpl(ruleValidator);
+        // trigger reload
+        // -> blocks classification requests until classification engine is ready
+        this.classificationEngine.reload();
```
Comment (fooker) [inline features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:97]: Initial reload should be done by `AsyncReloadingClassificationEngine` itself. This would make the wrapper more encapsulated.

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java [hunk 6]
```diff
@@ -322,7 +331,13 @@ public void validateRule(final Rule validateMe) {
         return sessionUtils.withTransaction(supplier);
     }
 
-    private void updateRulePositionsAndReloadEngine(final List<Rule> rules) {
+    private <T> T runInTransactionAndThenReload(Supplier<T> supplier) {
```
Comment (fooker) [inline features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:334]: While reading these changes, my initial thought was that `runInTransactionAndThenReload` will be extended by the async wrapper. This is clearly not the case. Therefore the whole renaming is purely cosmetic. Please avoid this kind of clutter.

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/decision/Tree.java [hunk 7]
```diff
@@ -452,11 +454,11 @@ public String toString() {
         }
 
         public final static class WithClassifiers extends Leaf {
-            public final List<Classifier> classifiers;
+            public final Classifiers classifiers;
 
             public WithClassifiers(List<Classifier> classifiers) {
                 super(Info.forLeaf(classifiers.size()));
-                this.classifiers = classifiers;
+                this.classifiers = new Classifiers(true, classifiers);
```
Comment (fooker) [inline features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/decision/Tree.java:461]: What does this do? Why is this change required?

**Review by fooker — CHANGES_REQUESTED** [review]
Please make sure that the commits message headers are correct before merging this (quash merging via the github UI and renaming there should be fine).

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (swachter) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:428]: renamed to `onReloadSucceded` and `onReloadFailed`

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/decision/Tree.java (hunk 7 above)
Comment (swachter) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/decision/Tree.java:461]: The change reduces allocations during classification.

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java [hunk 8]
```diff
@@ -358,4 +369,124 @@ private void checkForDuplicateName (Group group) {
...
+                    LOG.debug("another classification engine reload is required");
+                    doReload();
+                } else {
+                    LOG.debug("classification engine reloaded");
+                }
+            }
+        }
+
+        private synchronized boolean reloadSucceeded() {
+            if (state == State.NEED_ANOTHER_RELOAD) {
```
[the remark this answers is not in the document]
Comment (swachter) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:446]: The delegate has no state of its own. The delegate is always ready to do classifications. When `reloadSucceded` is called no other reload is in progress. (Ignoring possible background reload triggers; but these have no impact because reloads are idempotent.)

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java (hunk 4 above)
Comment (swachter) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:90]: I think it is a matter of taste. The `AsyncReloadingClassificationEngine` was only introduced to support consistent and non-sluggish front-end behavior. It allows to repeatedly trigger reloads but blocks on classification if a reload is on-going.

The other classification engine instances that are configured in the application context (`DefaultClassificationEngine` and `TimingClassificationEngine`) are of general use. In particular, these two classification are non-blocking even in case of on-going reloads.

The `AsyncReloadingClassificationEngine` is only used in the `DefaultClassificationService` that underlies the `ClassificationRestService`.

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java (hunk 5 above)
Comment (swachter) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:97]: done

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java (hunk 6 above)
Comment (swachter) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:334]: I disagree with your judgement. `runInTransactionAndThenReload` combines two steps that are executed at several places. The "reload" was indicated in method names earlier. Therefore I added it as a suffix to the new method.

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java (hunk 4 above)
Comment (fooker) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:90]: It's currently only used that way. Please do note remove the flexibility we head before.

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java (hunk 6 above)
Comment (fooker) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:331]: But why have you moved the reload from these methods in the first place? Just to get rid of the double-reload in `updateGroup`?

Comment (swachter) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java:331]: Yes, reload takes several seconds. Classification would be blocked 2 times as long.

**Review by cgorantla — APPROVED** [review]
LGTM!

File: features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java (hunk 4 above)
Comment (swachter) [reply features/flows/classification/engine/impl/src/main/java/org/opennms/netmgt/flows/classification/internal/DefaultClassificationService.java]: AsyncReloadingClassification is now wired in app context.

**Review by fooker — APPROVED** [review]
Please make sure the commits are squashed and named properly.
